### ISSUE TITLE
[bug] WITH_AVX not honored

### ISSUE THREAD

**Issue by roman-orekhov** [xiph/flac#286]
CMake option `WITH_AVX` translates to `FLAC__USE_AVX` which is only [used for GCC 4.9+](https://github.com/xiph/flac/blob/master/src/libFLAC/include/private/cpu.h#L108), but that `#ifdef` is always true by [definition](https://github.com/xiph/flac/blob/master/config.cmake.h.in#L48) of `FLAC__USE_AVX`.

It should be used for every compiler instead and as `#if FLAC__USE_AVX`

**Comment by NotTsunami** [comment]
Just for a point of reference for those concerned, the configurability for CMake was removed in ce6dd6b5732e319ef60716d9cc9af6a836a4011a. The linked pull request fixes it.

**Comment by ebachard** [comment]
Hello,

Thanks for the first changes, very helpfull. FYI, under Linux, I had to cross compile libflac for Windows (7+) using mingw **without avx** and I was able to build the current git version without issue. 

But it does not work : the reason is that avx2 .c files are yet compiled, and there are still avx2 symbols in libflac.a at the end (unexpected)

As workaround, I propose the following changes. If you are interested I can create a PR and add the license you want. 

Last but not least,  egrep -Hr  avx*  in libflac.a and any installed libs do no longer contain avx symbols nor avx assembly. 


````
--- a/src/libFLAC/CMakeLists.txt
+++ b/src/libFLAC/CMakeLists.txt
@@ -46,7 +46,7 @@ endif()
 
 include_directories("include")
 
-add_library(FLAC
+set(FLAC_SOURCES
     bitmath.c
     bitreader.c
     bitwriter.c
@@ -60,7 +60,6 @@ add_library(FLAC
     lpc.c
     lpc_intrin_sse2.c
     lpc_intrin_sse41.c
-    lpc_intrin_avx2.c
     lpc_intrin_vsx.c
     md5.c
     memory.c
@@ -70,7 +69,6 @@ add_library(FLAC
     stream_encoder.c
     stream_encoder_intrin_sse2.c
     stream_encoder_intrin_ssse3.c
-    stream_encoder_intrin_avx2.c
     stream_encoder_framing.c
     window.c
     $<$<BOOL:${WIN32}>:../../include/share/windows_unicode_filenames.h>
@@ -79,6 +77,16 @@ add_library(FLAC
     $<$<BOOL:${OGG_FOUND}>:ogg_encoder_aspect.c>
     $<$<BOOL:${OGG_FOUND}>:ogg_helper.c>
     $<$<BOOL:${OGG_FOUND}>:ogg_mapping.c>)
+
+if(WITH_AVX)
+set(FLAC_SOURCES
+    ${FLAC_SOURCES}
+    lpc_intrin_avx2.c
+    stream_encoder_intrin_avx2.c)
+endif()
+
+add_library(FLAC ${FLAC_SOURCES})
+
 set_property(TARGET FLAC PROPERTY PROJECT_LABEL "libFLAC")
 if(TARGET FLAC-asm)
     target_sources(FLAC PRIVATE $<TARGET_OBJECTS:FLAC-asm>)

````

**Comment by ktmf01** [comment]
@ebachard have you tried the fix under PR #299 or did you build from master?

**Comment by ebachard** [comment]
Apologies, I didn't see you answered me already.

Using master : git  b171e767e282909d3f4ef4b7d7fcb9b7d7987ff5
````
$ git log 
commit b171e767e282909d3f4ef4b7d7fcb9b7d7987ff5 (HEAD -> master, origin/master, origin/HEAD)
Author: NotTsunami <4589807+NotTsunami@users.noreply.github.com>
Date:   Mon Apr 18 02:45:14 2022 -0400

    Add missing langinfo check for CMake
....
````


First step : I tested all the changes, including the one in #299. Note: I can provide you the full diff (just in case)

Second step : 
````
  nm -a libflac.a | grep 
````
told me several labels containing avx2 were inside, and I was really not sure #299 was really working.

Step 3 : after reading the code, it become obvious to me to not compile lpc_intrin_avx2.c and    stream_encoder_intrin_avx2.c 
and I simply modified the cmakefile to not include them in the list when avx instructions are not expected.

Step 4 : no more avx*  thing'ry in libflac.a at the end => my fix work, but I'm unsure it is correct.

No answer issue with libflac tough, and if you are interested, I can provide you more information.

Thanks for your support :-)

**Comment by ebachard** [comment]
To avoid misunderstanding, here are all my changes (including #299 one)

[all_changes_build_without_avx2.diff.txt](https://github.com/xiph/flac/files/8543663/all_changes_build_without_avx2.diff.txt)

**Comment by ktmf01** [comment]
I just tested myself

With this patch, not disabling AVX
```
$ cmake ..; objdump -x src/libFLAC/libFLAC.a | grep avx
lpc_intrin_avx2.c.o:     file format elf64-x86-64
rw-r--r-- 0/0  28400 Jan  1 01:00 1970 lpc_intrin_avx2.c.o
0000000000000000 l    df *ABS*	0000000000000000 lpc_intrin_avx2.c
0000000000000000 l     F .text.unlikely	0000000000000008 FLAC__lpc_compute_residual_from_qlp_coefficients_16_intrin_avx2.cold
0000000000000008 l     F .text.unlikely	0000000000000008 FLAC__lpc_compute_residual_from_qlp_coefficients_intrin_avx2.cold
0000000000000010 l     F .text.unlikely	0000000000000007 FLAC__lpc_compute_residual_from_qlp_coefficients_wide_intrin_avx2.cold
0000000000000000 g     F .text	0000000000002053 FLAC__lpc_compute_residual_from_qlp_coefficients_16_intrin_avx2
0000000000002060 g     F .text	0000000000001d90 FLAC__lpc_compute_residual_from_qlp_coefficients_intrin_avx2
0000000000003df0 g     F .text	00000000000018fb FLAC__lpc_compute_residual_from_qlp_coefficients_wide_intrin_avx2
0000000000000000         *UND*	0000000000000000 FLAC__lpc_compute_residual_from_qlp_coefficients_intrin_avx2
0000000000000000         *UND*	0000000000000000 FLAC__lpc_compute_residual_from_qlp_coefficients_16_intrin_avx2
0000000000000000         *UND*	0000000000000000 FLAC__lpc_compute_residual_from_qlp_coefficients_wide_intrin_avx2
0000000000000000         *UND*	0000000000000000 FLAC__precompute_partition_info_sums_intrin_avx2
0000000000005b8e R_X86_64_GOTPCREL  FLAC__lpc_compute_residual_from_qlp_coefficients_intrin_avx2-0x0000000000000004
0000000000005b95 R_X86_64_REX_GOTPCRELX  FLAC__lpc_compute_residual_from_qlp_coefficients_16_intrin_avx2-0x0000000000000004
0000000000005ba4 R_X86_64_GOTPCREL  FLAC__lpc_compute_residual_from_qlp_coefficients_wide_intrin_avx2-0x0000000000000004
0000000000005c06 R_X86_64_REX_GOTPCRELX  FLAC__precompute_partition_info_sums_intrin_avx2-0x0000000000000004
000000000000706e R_X86_64_GOTPCREL  FLAC__lpc_compute_residual_from_qlp_coefficients_intrin_avx2-0x0000000000000004
0000000000007075 R_X86_64_REX_GOTPCRELX  FLAC__lpc_compute_residual_from_qlp_coefficients_16_intrin_avx2-0x0000000000000004
0000000000007084 R_X86_64_GOTPCREL  FLAC__lpc_compute_residual_from_qlp_coefficients_wide_intrin_avx2-0x0000000000000004
00000000000070e8 R_X86_64_REX_GOTPCRELX  FLAC__precompute_partition_info_sums_intrin_avx2-0x0000000000000004
00000000000084a8 R_X86_64_GOTPCREL  FLAC__lpc_compute_residual_from_qlp_coefficients_intrin_avx2-0x0000000000000004
00000000000084af R_X86_64_REX_GOTPCRELX  FLAC__lpc_compute_residual_from_qlp_coefficients_16_intrin_avx2-0x0000000000000004
00000000000084be R_X86_64_GOTPCREL  FLAC__lpc_compute_residual_from_qlp_coefficients_wide_intrin_avx2-0x0000000000000004
000000000000851f R_X86_64_REX_GOTPCRELX  FLAC__precompute_partition_info_sums_intrin_avx2-0x0000000000000004
000000000000999d R_X86_64_GOTPCREL  FLAC__lpc_compute_residual_from_qlp_coefficients_intrin_avx2-0x0000000000000004
00000000000099a4 R_X86_64_REX_GOTPCRELX  FLAC__lpc_compute_residual_from_qlp_coefficients_16_intrin_avx2-0x0000000000000004
00000000000099b3 R_X86_64_GOTPCREL  FLAC__lpc_compute_residual_from_qlp_coefficients_wide_intrin_avx2-0x0000000000000004
0000000000009a16 R_X86_64_REX_GOTPCRELX  FLAC__precompute_partition_info_sums_intrin_avx2-0x0000000000000004
000000000000ae42 R_X86_64_GOTPCREL  FLAC__lpc_compute_residual_from_qlp_coefficients_intrin_avx2-0x0000000000000004
000000000000ae49 R_X86_64_REX_GOTPCRELX  FLAC__lpc_compute_residual_from_qlp_coefficients_16_intrin_avx2-0x0000000000000004
000000000000ae58 R_X86_64_GOTPCREL  FLAC__lpc_compute_residual_from_qlp_coefficients_wide_intrin_avx2-0x0000000000000004
000000000000aeb7 R_X86_64_REX_GOTPCRELX  FLAC__precompute_partition_info_sums_intrin_avx2-0x0000000000000004
000000000000c386 R_X86_64_GOTPCREL  FLAC__lpc_compute_residual_from_qlp_coefficients_intrin_avx2-0x0000000000000004
000000000000c38d R_X86_64_REX_GOTPCRELX  FLAC__lpc_compute_residual_from_qlp_coefficients_16_intrin_avx2-0x0000000000000004
000000000000c39c R_X86_64_GOTPCREL  FLAC__lpc_compute_residual_from_qlp_coefficients_wide_intrin_avx2-0x0000000000000004
000000000000c3fd R_X86_64_REX_GOTPCRELX  FLAC__precompute_partition_info_sums_intrin_avx2-0x0000000000000004
stream_encoder_intrin_avx2.c.o:     file format elf64-x86-64
rw-r--r-- 0/0   4392 Jan  1 01:00 1970 stream_encoder_intrin_avx2.c.o
0000000000000000 l    df *ABS*	0000000000000000 stream_encoder_intrin_avx2.c
0000000000000000 g     F .text	0000000000000b8e FLAC__precompute_partition_info_sums_intrin_avx2
```

With this patch, disabling AVX
```
$ cmake .. -DWITH_AVX=OFF; objdump -x src/libFLAC/libFLAC.a | grep avx
lpc_intrin_avx2.c.o:     file format elf64-x86-64
rw-r--r-- 0/0   1088 Jan  1 01:00 1970 lpc_intrin_avx2.c.o
0000000000000000 l    df *ABS*	0000000000000000 lpc_intrin_avx2.c
stream_encoder_intrin_avx2.c.o:     file format elf64-x86-64
rw-r--r-- 0/0   1096 Jan  1 01:00 1970 stream_encoder_intrin_avx2.c.o
0000000000000000 l    df *ABS*	0000000000000000 stream_encoder_intrin_avx2.c
```

Without this patch, disabling AVX returns exactly the same as the first output

So @ebachard, you are right that there are still AVX symbols in the output, but not of any functions. In the same fashion there are also VSX symbols in the output:

```
objdump -x src/libFLAC/libFLAC.a | grep vsx
lpc_intrin_vsx.c.o:     file format elf64-x86-64
rw-r--r-- 0/0   1088 Jan  1 01:00 1970 lpc_intrin_vsx.c.o
0000000000000000 l    df *ABS*	0000000000000000 lpc_intrin_vsx.c
```

**Comment by ebachard** [comment]
Hello,

Reading lpc_intrin_avx2.c, there are ~ 1000 lines of asm instructions, conditioned to FLAC_AVX2_SUPPORTED.
So there are two possibilities :
- either we add dead code (why compile c useless files and link them ?)
- or we add unexpected asm insctructions

But maybe I'm wrong somewhere ?

Thanks in advance for enligthning me :-)

**Comment by ktmf01** [comment]
I don't understand what you're asking. Could you rephrase? What is the problem you're experiencing?

Please note that what we're referring to in this issue is compile time selection of AVX2, but there is also a run-time selection on top of that. So, even if you select WITH_AVX, the resulting binary will still be able to run on non-AVX2 systems because of the run-time selection.
